extensions/amp-lightbox/1.0/lightbox.js

```diff
@@ -0,0 +1,196 @@
+  return new Promise(async (resolve, reject) => {
+    if (time <= 0) {
+      return reject(`${selector} not found`);
+    }
+    const elements = document.querySelectorAll(selector);
+    if (elements.length) {
+      return resolve(elements);
+    }
+    return setTimeout(async () => await getElement(selector, time - 1), 1000);
+  });
+}
+
+/**
+ * @param {!LightboxDef.Props} props
+ * @param {{current: (!LightboxDef.LightboxApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function LightboxWithRef(
+  {
+    id,
+    animateIn = 'fade-in',
+    closeButtonAriaLabel,
+    // eslint-disable-next-line no-unused-vars
+    scrollable, // (TODO: discussion)
+    children,
+    onOpen,
+    initialOpen,
+    ...rest
+  },
+  ref
+) {
+  const [show, setShow] = useState(initialOpen);
+  const [visible, setVisible] = useState(false);
+  const styleClasses = useStyles();
+  const onOpenFnRef = useRef();
+  onOpenFnRef.current = onOpen;
+  const animateInRef = useRef();
+  animateInRef.current = animateIn;
+  const lightboxContainWrapperRef = useRef();
+  lightboxContainWrapperRef.current = `.${styleClasses.lightboxContainWrapper}`;
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      open: () => {
+        setShow(true);
+      },
+      close: () => {
+        setShow(false);
+      },
+    }),
+    []
+  );
+
+  useEffect(() => {
+    if (show) {
+      setVisible(true);
+    } else {
+      getElement(lightboxContainWrapperRef.current)
+        .then((element) => {
+          element = element[0];
+          const animation = element
+            .animate(ANIMATION_PRESETS[animateInRef.current], {
+              duration: ANIMATION_DURATION,
+              direction: 'reverse',
+              fill: 'both',
+            })
+            .finished.then((opt_el) => {
+              setVisible(false);
+            })
+            .catch((err) => {
+              throw Error('Lightbox closing animation failed', err);
+            });
+          return () => animation.cancel();
+        })
+        .catch((err) => {
+          throw Error(err);
+        });
+    }
+  }, [show]);
+
+  useEffect(() => {
+    if (!show && !visible) {
+      return;
+    }
+    if (visible) {
+      getElement(lightboxContainWrapperRef.current)
+        .then((element) => {
+          element = element[0];
+          const animation = element.animate(
+            ANIMATION_PRESETS[animateInRef.current],
+            {
+              duration: ANIMATION_DURATION,
+              fill: 'both',
+            }
+          );
+          return () => animation.cancel();
+        })
+        .catch((error) => {
+          throw Error('Element not found:', error);
+        });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [visible]);
+
+  useLayoutEffect(() => {
+    const escFunction = (event) => {
+      if (event.keyCode === 27) {
+        setShow(false);
+      }
+    };
+    document.addEventListener('keydown', escFunction);
```

Comment:
Is there a way to use the lightbox element to listen to keydown event?

E.g.

```
return (show && (
   <div onKeyDown={...}
))
```

?

If not, then we need to do a couple of things here:
1. We need to find the correct document. E.g. `lightboxRef.current.ownerDocument`.
2. We need to ensure that this effect is removed when the lightbox is shown, otherwise it will keep handling this event, event when it's hidden.

Overall, for scoping, it'd be much cleaner to just set the onKeyDown on the lightbox element itself.